### PR TITLE
added pinned by correlation PS consumer strategy

### PR DESCRIPTION
**Pull request by sammosampson**
This pull request allows persistent subscriptions to have a consumer strategy that will present events to the same consumer per correlation id.

It uses the hash of the correlation property (specified by the "by_correlation_id" system projection) in meta data as the assignment to the consumer. This is very similar to the "Pinned" strategy except that it uses correlation id instead of the source stream, hence its name "PinnedByCorrelation". It even falls back to the "Pinned" strategy if no correlation exists in the metadata of an event.

I have added it as a system consumer strategy rather than a plugin, enabling it to share much of the implementation of the Pinned strategy. And so it is set up the same way as those. I also amended the app.min.js to include it in dropdowns in the Web UI. 

The main thing I wasn't quite sure of was how to pass the correlation id from The `ByCorrelation` projection manager to the plugin, so I ended up using a static prop on a static class `CorrelationPropertyIdContext`, so please let me know if this is satisfactory or not. 

Tests have been added also.

### CODE REVIEW

**Comment by gregoryyoung** [comment]
This seems like a legit strategy! Should be quite useful.

One thought. Might we want to create a byXstrategy and just have the varying ones all use it? Note this is a comment/question not a suggestion for this pr.

File: src/EventStore.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.26124.0
+# Visual Studio Version 16
```
Comment (pgermishuys) [inline src/EventStore.sln]: The changes in this file can safely be reverted.

File: src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js [hunk 2]
```diff
@@ -17,5 +17,5 @@ return this.EOL_CHAR!=e?(this.EOL_CHAR=e,!0):void 0},this.setPadding=function(e)
...itEnabled required class=form-table name="emit"></td></tr><tr><td>Track Emitted Streams</td><td><input type=checkbox ng-model=config.trackEmittedStreams required class=form-table name="trackEmittedStreams"></td></tr><tr><td>Checkpoint After (ms)</td><td><input ng-model=config.checkpointAfterMs required class=form-table name="checkpointAfterMs"></td></tr><tr><td>Checkpoint Handled Threshold</td><td><input ng-model=config.checkpointHandledThreshold required class=form-table name="checkpointHandledThreshold"></td></tr><tr><td>Checkpoint Unhandled Threshold (bytes)</td><td><input ng-model=config.checkpointUnhandledBytesThreshold required class=form-table name="checkpointUnhandledBytesThreshold"></td></tr><tr><td>Pending Events Threshold</td><td><input ng-model=config.pendingEventsThreshold required class=form-table name="pendingEventsThreshold"></td></tr><tr><td>Max Write Batch Length</td><td><input ng-model=config.maxWriteBatchLength required class=form-table name="maxWriteBatchLength"></td></tr><tr><td>Maximum Number Of Allowed Wries In Flight</td><td><input ng-model=config.maxAllowedWritesInFlight required class=form-table name="maxAllowedWritesInFlight"></td></tr></tbody></table>')
```
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: > Hey @sammosampson, thanks for the PR. In general (This hasn't been tested yet), it looks good, just a couple of minor things that we've noticed which is mentioned in the review.
> 
> The big one is the one that you've mentioned which is the static property. I was thinking of broadcasting the update of the projection correlation id property via a message and have the persistent subscription service handle this and update all the relevant subscriptions with this property. Unfortunately I don't see an easy way of doing that by having a quick look and with that being said, I think we are more than happy to accept this PR and fix this up later. What do you think of the suggested solution to the static property?

I think if you guys are happy with static property for now then fair enough. Message for doing this sounds like a good idea, but i dont really know  how i would go about implementing that either

[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: > For the UI, the changes could be made [here](https://github.com/EventStore/EventStore.UI/blob/master/src/js/modules/competing/controllers/SubscriptionsNewCtrl.js#L18.).

I'll do that forthwith

File: src/EventStore.ClusterNode/Properties/launchSettings.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode/Properties/launchSettings.json]: Will do

File: src/EventStore.sln (hunk 1 above)
Comment (sammosampson) [reply src/EventStore.sln]: On it

File: src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: Done

File: src/EventStore.sln (hunk 1 above)
Comment (sammosampson) [reply src/EventStore.sln]: Done

File: src/EventStore.ClusterNode/Properties/launchSettings.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode/Properties/launchSettings.json]: Done

File: src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: Do I need to revert the changes in app.min.js?

**Comment by sammosampson** [comment]
Did i break all the checks somehow?

**Comment by sammosampson** [comment]
Only mac os64 builds not working now guys, dont know why

**Comment by samholder** [comment]
@pgermishuys are there further changes required for this? And any idea about the cause of the build failure in the macx64 Release branch build?

**Comment by pgermishuys** [comment]
Hey @sammosampson and @samholder, my apologies. I was out of the office for a while. Will review the PR today.

File: src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgermishuys) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: Yes please @sammosampson.

File: src/EventStore.sln [hunk 4]
```diff
@@ -42,7 +42,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		Directory.Build.props = Directory.Build.props
 	EndProjectSection
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EventStore.Common.Utils", "EventStore.Common.Utils\EventStore.Common.Utils.csproj", "{647867C2-54FA-4C2C-9BCD-7F85A4A8F787}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EventStore.Common.Utils", "EventStore.Common.Utils\EventStore.Common.Utils.csproj", "{647867C2-54FA-4C2C-9BCD-7F85A4A8F787}"
```
Comment (pgermishuys) [inline src/EventStore.sln]: This change can also be reverted @sammosampson

**Comment by pgermishuys** [comment]
@samholder and @sammosampson, re-ran the builds on macos and they seem fine now. All the checks have passed. After the last change, it should be good to merge.

File: src/EventStore.sln (hunk 4 above)
Comment (sammosampson) [reply src/EventStore.sln]: Done

**Comment by sammosampson** [comment]
> @samholder and @sammosampson, re-ran the builds on macos and they seem fine now. All the checks have passed. After the last change, it should be good to merge.

All changes done now

File: src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sammosampson) [reply src/EventStore.ClusterNode.Web/clusternode-web/js/app.min.js]: done

**Pull request merged**